File: problem322.py
"""

                            11
                    /       |       \
                   6        9       10
                /   |   \
              1     4   5
             /|\
           -5-1 0


"""

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        result=float('inf')
        coins=sorted(coins, reverse=True)
        memoization = {0:0}
        def performChange(x):
            nonlocal coins, memoization
            if x==0:
                return 0
            if x<0:
                return float('inf')
            if x in memoization:
                return memoization[x]
            res = float('inf')
            for c in coins:
                res = min(res, 1+performChange(x-c))
            memoization[x]=res
            return res
        
        result=performChange(amount)
        
        if result==float('inf'):
            return -1
        return result
        
        
#Time: O(N*c),
#Space: O(N)

File: problem322.cs
 /*

 0 1 2
[5,2,1]


                            11
                    /       |
                    6     
            /       |       \
            1       4       5
    /   |  \    /   |   \
    x   x  0    x   2    3
            |
            3



    
*/

public class Solution {
    int[] dp;
    int[] coins;
    public int CoinChange(int[] coins, int amount) {
        this.coins = coins;
        this.dp = Enumerable.Repeat(Int32.MaxValue, amount+1).ToArray();
        dp[0]=0;
        
        for(int i=1; i<=amount; i++){
            for(int j=0; j<coins.Length; j++){
                if(i>=coins[j] && dp[i-coins[j]]!=Int32.MaxValue){
                    dp[i] = Math.Min(dp[i], dp[i-coins[j]]+1);
                }
            }
        }
        
        if(dp[amount]==Int32.MaxValue)
            return -1;
        
        return dp[amount];
    }
}

//time: O(N*Amount)
//Space: O(Amount)